### PR TITLE
[RFC] Replace Grantlee

### PR DESCRIPTION
**Pull request by atdotde**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

I was looking into our printing capabilities (which so far, I have never used) and stumbled getting Grantlee to work for me. Instead of wasting more time to get it to work I decided to reimplement the small subset of its features that we actually use directly.

Even if this code will not make it to master (which I still hope it will eventually do as I consider this a significant simplification), for me this was a nice exercise in implementing a parser for this very limited programming language.

Given that the amount of code to implement the whole thing is not much bigger than the code we used to call Grantlee I think this is a worthwhile simplification. The whole Grantlee library is far too heavy weight  and complex for those few replacements we actually use in our templates.

In the process, I also learned that our code actually circumvents QObject's introspection interface that Grantlee was supposed to make use of so using this heavy lifting wasn't really justified in the first place.

I made sure that at least those templates that we ship with the source work as before.

I have not yet started removing references to Grantlee from the build system(s) (including cmake) as this might be rather difficult given that I cannot test most of them.

### CODE REVIEW

File: desktop-widgets/templatelayout.cpp [hunk 1]
```diff
@@ -154,46 +149,25 @@ QString TemplateLayout::generate()
...
+	QHash<QString, QVariant> options;
+	options["print_options"] = QVariant::fromValue(*printOptions);
+	options["template_options"] = QVariant::fromValue(*templateOptions);
+	options["dives"] = QVariant::fromValue(diveList);
+	QList<token> tokens = lexer(templateContents);
+	QString buffer;
+	QTextStream out(&buffer);
+	int pos = 0;
+	parser(tokens, pos, out, options);
+	htmlContent = out.readAll();
```
[the remark this answers is not in the document]
Comment (atdotde) [reply desktop-widgets/templatelayout.cpp:164]: This is a left-over from the old code with the additional benefit that I could print it during debugging this (as you cannot puts(out.readAll()); return out.readAll(); as I can tell you from trying and wondering why nothing is returned ;-) )

**Comment by atdotde** [comment]
I understand that usually a lexer returns a promise. Here, I store the list of tokens as an intermediate value which I think is forgivable given the size of our template files. This is also because the result of the parser is not a syntax tree but it directly does the string replacement. The challenge is to handle (possibly nested) for-loops that need to repeat part of the token string, so it has to be stored anyway.

**Comment by dirkhh** [comment]
Sadly my migraines are back so instead of dealing with all the outstanding reviews and a few other loose ends I didn't do a thing on Subsurface over the weekend.
Assuming I'll feel better today, I'll try to find time to look into this some more.

**Comment by dirkhh** [comment]
So I spent a few minutes with this (without looking at the code, just trying to make sure it works at all).
A few questions:
- does this still require QtWebKit? I.e., am I correct to assume that this only drops the need for Grantlee, not the need for WebKit?
- Given that it should drop Grantlee, why is it still linking against it / complaining about metatypes?
- this warning indicates that there's something amiss with changes to the DPI value: `QMetaObject::connectSlotsByName: No matching signal for on_resolution_edited(int)` (again, I need to look at the code - we usually try to use the new typed `connect()` syntax, not the old name based one that gets evaluated at run time
- trying to print something with this branch gives me something very odd. A small rectangle in the top left corner of the page with a partial Subsurface logo and some of the keywords from the template I tried to use.
Screen Shot 2020-08-24 at 2 13 14 PM

**Comment by atdotde** [comment]
To your points:

- Yes, the WebKit dependency is (unfortunately) not affected. We still need WebKit's ability to access individual tags in the html to do the page breaking.
- I did not yet touch the build system (including cmake files). But it should be save to remove mentions of Grantlee from those. There shouldn't be complaints about metatypes. But still everything goes through QVariants and my understanding is that the metatypes are used to decode those.
- Complaints about on_resolution_edited are probably from the other patch about print resolutions. Will look into those.
- Could you email me the template you used. This is a case of ShouldNotHappen. Do you see this for all templates?

File: desktop-widgets/templatelayout.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bstoeger) [reply desktop-widgets/templatelayout.cpp:164]: Seems that my last comment was lost:
Why is it necessary to do that `readAll()`? Can't you simply access the backing buffer directly,
i.e. `return buffer`?

**Comment by bstoeger** [comment]
> This is also because the result of the parser is not a syntax tree but it directly does the string replacement.

In the long run, a parse tree would certainly be the preferred way to do things...

**Comment by dirkhh** [comment]
I would love to get back to this... so this still requires WebKit but gets rid of Grantlee.
Is there any specific functionality that should be tested at this stage?
Is there anything that is known to break?
I will go back to try this again and see if what I posted above was a bad build or is reproducible. Because now is the time to merge this with all the changes in flight - not closer to a release :-)

**Comment by atdotde** [comment]
This only replaces the template functions that fill in variables. It worked for me for the templates provided but I don't know if your problem of Aug 24 is still there. WebKit is still required for breaking the result into pages (as it allows to access html elements while in web engine this would have to be done through javascript).

So far, I had one unsuccessful attempt of actually removing mentions of Grantlee from the build system (in particular the cmake files) that was too aggressive in removing things (I remember there were some variables set in some included file that still needed to be there). I could give that another try. But help would be appreciated there since I am happy if building works on my Mac but I have never looked at any depth into the other platforms.

**Comment by dirkhh** [comment]
I'm planning to look into my template issues today. Removal from the buildsystem I can take care of.
No worries. But let's first check if this all works as advertised. :-)

Removing WebKit... yeah, that would be neat, but we need it for the user manual as well. And as much as I **HATE** having to build WebKit from source every time I move to a new Qt version... yeah, it's hard to rip out.

**Comment by glance-** [comment]
We had code which used webengine to render the manual https://github.com/subsurface/subsurface/commit/643f4a5726926b86693b442e318f40cf28fee04a

**Comment by dirkhh** [comment]
Fun. Or in other words... boy it's been a long time since I tried to print anything.
This seems to work just as well as the grantlee code for me. That said, either way I have some concerns - but those aren't for this PR, I think.
Quick look at these two excerpts from prints:

Screen Shot 2020-10-07 at 10 25 01 AM

having no default cylinder should not make us print a cylinder type of "no default cylinder"

Screen Shot 2020-10-07 at 10 24 47 AM

those are some pretty mighty gas change icons...
But I get AFAICT identical output with master - which why this isn't part of this PR review :-)

I did a limited amount of testing (three different templates) and it seems to all be fine.
Let me work on the build system changes and then maybe we can merge this...

**Comment by dirkhh** [comment]
and here we go with Grantlee gone from the build system -> #3007

**Comment by dirkhh** [comment]
merged via #3007

**Pull request closed without merging**